### ISSUE TITLE
Siamese MRCNN

### ISSUE THREAD

**Issue by aronfothi** [matterport/Mask_RCNN#1238]
Hi,
I would like to build a siamese mrcnn network where i copy the mrcnn model as the two branches of the my siamese model. I have to wrap the mrcnn model as a sub model into my new main model. 
I can do it in my toy model but when i try it with this implementation the sub model Inputs do not get value during the training

What i try:
`input_image = KL.Input(
                shape=[None, None, config.IMAGE_SHAPE[2]], name="input_image")
            input_image_meta = KL.Input(shape=[config.IMAGE_META_SIZE],
                                        name="input_image_meta")
            # RPN GT
            input_rpn_match = KL.Input(
                shape=[None, 1], name="input_rpn_match", dtype=tf.int32)
            input_rpn_bbox = KL.Input(
                shape=[None, 4], name="input_rpn_bbox", dtype=tf.float32)

            # Detection GT (class IDs, bounding boxes, and masks)
            # 1. GT Class IDs (zero padded)
            input_gt_class_ids = KL.Input(
                shape=[None], name="input_gt_class_ids", dtype=tf.int32)
            # 2. GT Boxes in pixels (zero padded)
            # [batch, MAX_GT_INSTANCES, (y1, x1, y2, x2)] in image coordinates
            input_gt_boxes = KL.Input(
                shape=[None, 4], name="input_gt_boxes", dtype=tf.float32)
            
            # 3. GT Masks (zero padded)
            # [batch, height, width, MAX_GT_INSTANCES]
            if config.USE_MINI_MASK:
                input_gt_masks = KL.Input(
                    shape=[config.MINI_MASK_SHAPE[0],
                        config.MINI_MASK_SHAPE[1], None],
                    name="input_gt_masks", dtype=bool)
            else:
                input_gt_masks = KL.Input(
                    shape=[config.IMAGE_SHAPE[0], config.IMAGE_SHAPE[1], None],
                    name="input_gt_masks", dtype=bool)

            input_instance_id = KL.Input(
                    shape=[config.IMAGE_SHAPE[0], config.IMAGE_SHAPE[1]],
                    name="input_instance_id", dtype=tf.int32)

            inputs = [input_image, input_image_meta,
                        input_rpn_match, input_rpn_bbox, input_gt_class_ids, input_gt_boxes, input_gt_masks, input_instance_id]
            if not config.USE_RPN_ROIS:
                inputs.append(input_rois)

            sub_model = self.build_sub_model(config)

            outputs = sub_model(inputs)
            
            
            model = KM.Model(inputs, outputs, name='siamese')
            model.get_layer("mask_rcnn").inputs = model.inputs
            print('model', model.summary())
            print(model.inputs, model.get_layer("mask_rcnn").inputs)
            return model`

where build_sub_model is the original build
I got this error message:
InvalidArgumentError (see above for traceback): You must feed a value for placeholder tensor 'input_gt_masks_sub' with dtype bool and shape [?,56,56,?]
	 [[node input_gt_masks_sub (defined at /home/fothar/.local/lib/python3.5/site-packages/keras/backend/tensorflow_backend.py:504)  = Placeholder[dtype=DT_BOOL, shape=[?,56,56,?], _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]

My toy model that works:

`from __future__ import print_function
import keras
from keras.datasets import mnist
from keras.models import Sequential, Model
from keras.layers import *
from keras import backend as K
from keras.callbacks import TensorBoard

batch_size = 128
num_classes = 10
epochs = 12

# input image dimensions
img_rows, img_cols = 28, 28

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

if K.image_data_format() == 'channels_first':
    x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
    x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
    input_shape = (1, img_rows, img_cols)
else:
    x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
    input_shape = (img_rows, img_cols, 1)

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)

def build_sub_model(input_shape):

    #model = Sequential()
    inpa = Input(input_shape, name="sub_inpa")
    inpb = Input(input_shape, name="sub_inpb")
    
    x = Conv2D(32, kernel_size=(3, 3),
                    activation='relu')(inpa)
    x = Conv2D(64, (3, 3), activation='relu')(x)
    x = MaxPooling2D(pool_size=(2, 2))(x)
    x = Dropout(0.25)(x)
    x = Flatten()(x)
    x = Dense(128, activation='relu')(x)
    x = Dropout(0.5)(x)
    x = Dense(num_classes, activation='softmax')(x)
    model = Model([inpa, inpb], [x, inpb], name="sub_model")
    print(model.summary())
    print(model.inputs)
    return model

def build_model():

    sub_model = build_sub_model(input_shape)

    inp2a = Input(input_shape, name="inp2a")
    inp2b = Input(input_shape, name="inp2b")

    inputs = [inp2a, inp2b]
    sub_out = sub_model(inputs)
    #print(len(x_inpb), x_inpb)
    model = Model(inputs, sub_out)
    print(model.summary())
    print(model.get_layer("sub_model").inputs)
    model.get_layer("sub_model").inputs = model.inputs
    print(model.summary())
    print(model.get_layer("sub_model").summary())
    print(model.get_layer("sub_model").inputs)
    for i in model.get_layer("sub_model").layers:
        if isinstance(i, InputLayer):
            print('sub', i, i.name, i.get_output_at(0))
    print(model.inputs)
    for i in model.layers:
        if isinstance(i, InputLayer):
            print('main', i, i.name, i.get_output_at(0))
    
    return model

model2 =  build_model()

model2.compile(loss=[keras.losses.categorical_crossentropy, keras.losses.mean_squared_error],
               loss_weights=[1., 1.],
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])
callbacks = [
             TensorBoard(log_dir='/home/fothar/cluster_segmentation_redesign/logs/xxx/',
                                        histogram_freq=0, write_graph=True, write_images=False)
        ]              

model2.fit([x_train, x_train], [y_train, x_train],
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          callbacks=callbacks,
          validation_data=([x_test, x_test], [y_test, x_test]))
score = model2.evaluate([x_test, x_test], [y_test, x_test], verbose=0)
print(len(score), score)`
